flink-end-to-end-tests/flink-metrics-reporter-prometheus-test/src/test/java/org/apache/flink/metrics/prometheus/tests/PrometheusReporterEndToEndITCase.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.TestLogger;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+
+import okhttp3.OkHttpClient;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.tests.util.AutoClosableProcess.runBlocking;
+import static org.apache.flink.tests.util.AutoClosableProcess.runNonBlocking;
+
+/**
+ * End-to-end test for the PrometheusReporter.
+ */
+public class PrometheusReporterEndToEndITCase extends TestLogger {
+
+	private static final Logger LOG = LoggerFactory.getLogger(PrometheusReporterEndToEndITCase.class);
+
+	private static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
+
+	private static final String PROMETHEUS_FILE_NAME = "prometheus-2.4.3.linux-amd64";
+
+	private static final Pattern LOG_REPORTER_PORT_PATTERN = Pattern.compile(".*Started PrometheusReporter HTTP server on port ([0-9]+).*");
+
+	@Rule
+	public final FlinkDistribution dist = new FlinkDistribution();
+
+	@Rule
+	public final TemporaryFolder tmp = new TemporaryFolder();
+
+	@Test
+	public void testReporter() throws Exception {
+		LOG.info("starting test");
+		dist.copyOptJarsToLib("flink-metrics-prometheus");
+
+		final Configuration config = new Configuration();
+		config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "prom." + ConfigConstants.METRICS_REPORTER_CLASS_SUFFIX, PrometheusReporter.class.getCanonicalName());
+		config.setString(ConfigConstants.METRICS_REPORTER_PREFIX + "prom.port", "9000-9100");
+
+		dist.appendConfiguration(config);
+
+		final Path tmpPrometheusDir = tmp.newFolder().toPath().resolve("prometheus");
+		final Path prometheusArchive = tmpPrometheusDir.resolve(PROMETHEUS_FILE_NAME + ".tar.gz");
+		final Path prometheusBinDir = tmpPrometheusDir.resolve(PROMETHEUS_FILE_NAME);
+		final Path prometheusConfig = prometheusBinDir.resolve("prometheus.yml");
+		final Path prometheusBinary = prometheusBinDir.resolve("prometheus");
+		Files.createDirectory(tmpPrometheusDir);
+
+		runBlocking(
+			"Download of Prometheus",
+			Duration.ofMinutes(5),
+			CommandLineWrapper
+				.wget("https://github.com/prometheus/prometheus/releases/download/v2.4.3/" + prometheusArchive.getFileName())
+				.targetDir(tmpPrometheusDir)
+				.build());
+
+		runBlocking("Extraction of Prometheus archive",
+			CommandLineWrapper
+				.tar(prometheusArchive)
+				.extract()
+				.zipped()
+				.targetDir(tmpPrometheusDir)
+				.build());
+
+		runBlocking("Set Prometheus scrape interval",
+			CommandLineWrapper
+				.sed("s/\\(scrape_interval:\\).*/\\1 1s/", prometheusConfig)
```

Comment:
Can't we implement behaviors like `sed` in java, otherwise we will be still tackling issues of platform dependent semantics of those tools.

Reply:
Of course we could, but I didn't want to do this _now_.

Reply:
I had already implemented the sed logic in bash (actually the entire test was already written in bash), so I opted for re-using some stuff.

Reply:
huh, I thought i mentioned this in the PR description but it must've been lost somewhere :/

Reply:
got it, thanks for explanation